### DISCUSSION TITLE
TCP ACKs Unrecognized on VICE C128 RR-Net

### DISCUSSION THREAD

**Discussion by defc0n** [cc65/cc65#2161] [Q&A] [answered]
I have the following code: https://pastebin.com/kBky6Nug and when I run it in VICE the following happens: https://imgur.com/a/GrlP96U

The pcap shows that the DNS query appears to work and that it tries to make a TCP connection to example.com's IP address. I see the incoming ACKs, but it seems the ip65 TCP stack isn't getting them.

I haven't tried this on real hardware (c128 w/ 64nic+) yet.

I also tried the example here https://github.com/cc65/ip65/blob/main/test/tcp.c and appear to be getting a timeout there as well.

I believe I'm using the latest cc65 & ip65 libraries.

**Comment by mrdudz** [comment]
have you verified it works *at all* in VICE? (eg use the precompiled ntp example for this)

**Reply by defc0n** [reply]
Hello @mrdudz ,

The precompiled DATE65 UDP program appears to be working fine in VICE C128 in C64 mode:
 
![Screenshot from 2023-08-19 11-48-21](https://github.com/cc65/cc65/assets/5515404/ff9d5d13-bedd-49c2-9841-c9a61865c5e1)

**Reply by defc0n** [reply]
Interestingly, the precompiled TELNET65 program is working fine too. So I'd think either something with my code is wrong or there is some issues with the libraries I'm using or the way I'm compiling it?

**Comment by mrdudz** [comment]
I'm not really familiar with that lib... but if the examples work, then perhaps @oliverschmidt can take a peek at your code? :)

**Comment by oliverschmidt** [comment]
> I also tried the example here https://github.com/cc65/ip65/blob/main/test/tcp.c and appear to be getting a timeout there as well.

Works for me:
![image](https://github.com/cc65/cc65/assets/2664009/ee475d9f-62d1-42f0-a00f-7757f03840f4)

**Comment by defc0n** [comment]
@oliverschmidt strange. I'm compiling the TCP program by running:
```
$ cl65 -o tcp.prg -O -t c64 test/tcp.c ip65.lib ip65_tcp.lib ip65_c64.lib
```
and modified it to connect to batmud.bat.org (95.175.124.84) port 23:
```
$ git diff test/tcp.c
diff --git a/test/tcp.c b/test/tcp.c
index 63a7a36..6a129b4 100644
--- a/test/tcp.c
+++ b/test/tcp.c
@@ -7,7 +7,7 @@
 #include "../inc/ip65.h"
 
 #define LEN 500
-#define SRV "192.168.0.10"
+#define SRV "95.175.124.84"
 
 unsigned char buf[1500];
 int len;
@@ -65,7 +65,7 @@ void main(void)
   printf("DNS Srv: %s\n", dotted_quad(cfg_dns));
 
   printf("Connect to %s\n", SRV);
-  if (tcp_connect(srv, 6502, tcp_recv))
+  if (tcp_connect(srv, 23, tcp_recv))
   {
     error_exit();
   }
```
But get:
![Screenshot from 2023-08-20 19-24-12](https://github.com/cc65/cc65/assets/5515404/7532d352-b3f5-45fd-9ef5-de954a4c75a7)

Yet when I run the `apps/telnet65.prg` it can connect fine:
![Screenshot from 2023-08-20 19-29-56](https://github.com/cc65/cc65/assets/5515404/8262664b-b5ae-460e-b0af-a33de25a2818)

So my TCP issues for things I'm trying to compile happen in c64 along with c128 it seems. I'm currently in the ip65 commit:
```
commit 821a63c0c219c0365bdfe58ac6699c9d9a2423e3 (HEAD -> main, tag: 2023-06-11, origin/main, origin/HEAD)
Author: Oliver Schmidt <ol.sc@web.de>
Date:   Sun Jun 11 12:23:45 2023 +0200

    Adjusted to cc65/cc65#1844

```

**Comment by oliverschmidt** [comment, marked as answer]
> strange.

Not at all.

> `$ cl65 -o tcp.prg -O -t c64 test/tcp.c ip65.lib ip65_tcp.lib ip65_c64.lib`

You have to link __EITHER__ ip65.lib __OR__ ip65_tcp.lib.

See https://github.com/cc65/ip65/wiki/Library-for-C-programs or https://github.com/cc65/ip65/blob/main/test/Makefile#L59-L60.

**Reply by defc0n** [reply]
This indeed appears to have fixed the issue, thanks.
